File: modyn/NewStorage/src/internal/file_wrapper/binary_file_wrapper.cpp
#include "internal/file_wrapper/binary_file_wrapper.hpp"

#include <fstream>
#include <numeric>
#include <vector>

using namespace storage;

int64_t BinaryFileWrapper::int_from_bytes(const unsigned char* begin, const unsigned char* end) {
  int64_t value = 0;
#if __BYTE_ORDER__ == __ORDER_BIG_ENDIAN__
  value = std::accumulate(begin, end, 0, [](uint64_t acc, unsigned char other) { return (acc << 8u) | other; });
#elif __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__
  value = std::accumulate(begin, end, 0, [](uint64_t acc, unsigned char other) { return (acc << 8u) | other; });
#else
#error "Unknown byte order"
#endif
  return value;
}

int64_t BinaryFileWrapper::get_number_of_samples() { return file_size_ / record_size_; }

void BinaryFileWrapper::validate_file_extension() {
  const std::string extension = file_path_.substr(file_path_.find_last_of('.') + 1);
  if (extension != "bin") {
    throw std::invalid_argument("Binary file wrapper only supports .bin files.");
  }
}

int64_t BinaryFileWrapper::get_label(int64_t index) {
  const int64_t record_start = index * record_size_;
  std::vector<unsigned char> data_vec = filesystem_wrapper_->get(file_path_);
  unsigned char* data = data_vec.data();
  unsigned char* label_begin = data + record_start;
  unsigned char* label_end = label_begin + label_size_;
  return int_from_bytes(label_begin, label_end);
}

std::vector<int64_t> BinaryFileWrapper::get_all_labels() {
  const int64_t num_samples = get_number_of_samples();
  std::vector<int64_t> labels = std::vector<int64_t>();
  labels.reserve(num_samples);
  std::vector<unsigned char> data_vec = filesystem_wrapper_->get(file_path_);
  unsigned char* data = data_vec.data();
  for (int64_t i = 0; i < num_samples; i++) {
    unsigned char* label_begin = data + (i * record_size_);
    unsigned char* label_end = label_begin + label_size_;
    labels.push_back(int_from_bytes(label_begin, label_end));
  }
  return labels;
}

std::vector<std::vector<unsigned char>> BinaryFileWrapper::get_samples(int64_t start, int64_t end) {
  const std::vector<int64_t> indices = {start, end};
  BinaryFileWrapper::validate_request_indices(get_number_of_samples(), indices);
  const int64_t num_samples = end - start + 1;
  const int64_t record_start = start * record_size_;
  const int64_t record_end = record_start + num_samples * record_size_;
  std::vector<unsigned char> data_vec = filesystem_wrapper_->get(file_path_);
  unsigned char* data = data_vec.data();
  std::vector<std::vector<unsigned char>> samples = std::vector<std::vector<unsigned char>>(num_samples);
  for (int64_t i = record_start; i < record_end; i += record_size_) {
    unsigned char* sample_begin = data + i + label_size_;
    unsigned char* sample_end = sample_begin + sample_size_;
    const std::vector<unsigned char> sample(sample_begin, sample_end);
    samples[(i - record_start) / record_size_] = sample;
  }
  return samples;
}

std::vector<unsigned char> BinaryFileWrapper::get_sample(int64_t index) {
  const std::vector<int64_t> indices = {index};
  BinaryFileWrapper::validate_request_indices(get_number_of_samples(), indices);
  const int64_t record_start = index * record_size_;
  std::vector<unsigned char> data_vec = filesystem_wrapper_->get(file_path_);
  unsigned char* data = data_vec.data();
  unsigned char* sample_begin = data + record_start + label_size_;
  unsigned char* sample_end = sample_begin + sample_size_;
  return {sample_begin, sample_end};
}

std::vector<std::vector<unsigned char>> BinaryFileWrapper::get_samples_from_indices(
    const std::vector<int64_t>& indices) {
  BinaryFileWrapper::validate_request_indices(get_number_of_samples(), indices);
  std::vector<std::vector<unsigned char>> samples = std::vector<std::vector<unsigned char>>();
  samples.reserve(indices.size());
  std::vector<unsigned char> data_vec = filesystem_wrapper_->get(file_path_);
  unsigned char* data = data_vec.data();
  for (const int64_t index : indices) {
    const int64_t record_start = index * record_size_;
    unsigned char* sample_begin = data + record_start + label_size_;
    unsigned char* sample_end = sample_begin + sample_size_;
    const std::vector<unsigned char> sample(sample_begin, sample_end);
    samples.push_back(sample);
  }
  return samples;
}

File: modyn/trainer_server/internal/dataset/binary_file_wrapper.py
"""Binary file wrapper."""
import os


class BinaryFileWrapper:
    """Binary file wrapper.

    Binary files store raw sample data in a row-oriented format. One file can contain multiple samples.
    This wrapper requires that each samples should start with the label followed by its set of features.
    Each sample should also have a fixed overall width (in bytes) and a fixed width for the label,
    both of which should be provided in the config. The file wrapper is able to read samples by
    offsetting the required number of bytes.
    """

    def __init__(self, file_path: str, byteorder: str, record_size: int, label_size: int):
        """Init binary file wrapper.

        Args:
            file_path (str): Path to file
            file_wrapper_config (dict): File wrapper config
            filesystem_wrapper (AbstractFileSystemWrapper): File system wrapper to abstract storage of the file

        Raises:
            ValueError: If the file has the wrong file extension
            ValueError: If the file does not contain an exact number of samples of given size
        """
        self.byteorder = byteorder
        self.file_path = file_path

        self.record_size = record_size
        self.label_size = label_size
        if self.record_size - self.label_size < 1:
            raise ValueError("Each record must have at least 1 byte of data other than the label.")

        self.file_size = os.path.getsize(self.file_path)

        if self.file_size % self.record_size != 0:
            raise ValueError("File does not contain exact number of records of size " + str(self.record_size))

    def get_number_of_samples(self) -> int:
        """Get number of samples in file.

        Returns:
            int: Number of samples in file
        """
        return int(self.file_size / self.record_size)

    def get_all_labels(self) -> list[int]:
        with open(self.file_path, "rb") as file:
            data = file.read()

        num_samples = self.get_number_of_samples()
        labels = [
            int.from_bytes(
                data[(idx * self.record_size) : (idx * self.record_size) + self.label_size], byteorder=self.byteorder
            )
            for idx in range(num_samples)
        ]
        return labels

    def get_sample(self, index: int) -> bytes:
        """Get the sample at the given index.
        The indices are zero based.

        Args:
            index (int): Index

        Raises:
            IndexError: If the index is out of bounds

        Returns:
            bytes: Sample
        """
        return self.get_samples_from_indices([index])[0]

    def get_samples(self, start: int, end: int) -> list[bytes]:
        return self.get_samples_from_indices(list(range(start, end)))

    def get_samples_from_indices(self, indices: list) -> list[bytes]:
        with open(self.file_path, "rb") as file:
            data = file.read()

        samples = [data[(idx * self.record_size) + self.label_size : (idx + 1) * self.record_size] for idx in indices]
        return samples
